### DISCUSSION TITLE
Request an example that captures the pattern of malloc and free matches in C

### DISCUSSION THREAD

**Discussion by fogcell** [ast-grep/ast-grep#885] [Q&A]
Request an example that captures the pattern of malloc and free matches in C I read the language example, but since I don't know the other languages well, I can't write a function call that matches pairs of matches.

For example,Can recognize that free(abc) was not called on return.My purpose is to use ast-grep to do a simple linter, such a requirement ast-grep implementation trouble?

```c
void func1()
{
  abc = malloc()；
 
  if(...)
  {
     return;
  } 


  free（abc）;
  return;
}
```
Another question:
1. look up malloc_abc() malloc_efg()?
2. Or provide some other knowledge points that I can learn on my own（Do I need to have AST knowledge?）

**Comment by HerringtonDarkholme** [comment]
Thanks for trying out ast-grep. My two suggestions:

1. Check out the the ast-grep official document and see how it works and what it can do.
2. refine you problem statement. finding malloc/free is hard and you probably want to support only a subset of condition. see https://reverseengineering.stackexchange.com/questions/11149/locating-memory-allocation-functions-from-static-analysis

**Reply by fogcell** [reply]
Thank you for replying to me so quickly, pleasantly, in my expectation

looking at previous discussions and similar discussion topics, I see that some of the questioners may be attempting to solve certain problems in a way that is not supported by ast-grep, so I wanted to see if there was a similar need I had that could be answered without me understanding it thoroughly enough. 

I checked your link and the questions and answers therein are not relevant to my intent.

Let me try to clarify my thoughts further：
malloc is just an example, I saw that ast-grep can be used as a linter, so I thought of this example, after calling malloc in the same function, according to normal behavior, you must call a free before returning to ensure a peer match.

ps. Maybe I really need to take a closer look at the ast-grep documentation, which has too few examples of C😢
